plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/repository/AbstractSourceAttachmentDownloaderJob.java

```diff
@@ -0,0 +1,142 @@
+    super("TestSourceAttachmentDownloaderJob");
+    this.javaProject = javaProject;
+    this.classpathEntryPath = classpathEntryPath;
+    this.serializer = serializer;
+    this.retryAttempts = retryAttempts;
+    this.retryDelayFactor = retryDelayFactor;
+  }
+
+  @Override
+  protected IStatus run(IProgressMonitor monitor) {
+    try {
+      IPath sourcePath = getSourcePath();
+      if (sourcePath != null) {
+        setSourceAttachmentPath(getJavaProject(), sourcePath, monitor);
+      }
+      return Status.OK_STATUS;
+    } catch (IOException | CoreException ex) {
+      return StatusUtil.error(this, "Could not attach source path", ex);
+    }
+  }
+
+  protected abstract IPath getSourcePath();
+
+  /**
+   * Finds the library entry with path matching {@link #classpathEntryPath} and sets the source attachment path to
+   * <code>sourcePath</code>
+   */
+  private void setSourceAttachmentPath(IJavaProject javaProject, IPath sourcePath, IProgressMonitor monitor) throws IOException, CoreException {
+    IClasspathEntry[] rawClasspath = javaProject.getRawClasspath();
+    for (int i = 0; i < rawClasspath.length; i++) {
+      // this method can return null for uninitialized LibraryClasspathContainer instances, in that case
+      // this jobs gets rescheduled to let JDT initialize the container in the meantime
+      LibraryClasspathContainer libraryContainer = getLibraryClasspathContainer(rawClasspath[i]);
+      if (libraryContainer != null) {
+        IClasspathEntry[] classpathEntries = libraryContainer.getClasspathEntries();
+        for (int j = 0; j < classpathEntries.length; j++) {
+          if (classpathEntries[j].getPath().equals(classpathEntryPath)) {
+            setSourceAttachmentPathForEntry(javaProject, sourcePath, monitor, libraryContainer, classpathEntries, j);
+            return;
+          }
+        }
+      }
+    }
+    if (retryAttempts-- > 0) {
+      schedule((long) (Math.random() * retryDelayFactor));
+    } else {
+      new CoreException(StatusUtil.error(this, "Could not set source attachment path"));
+    }
+  }
+
+  /**
+   * Returns the {@link LibraryClasspathContainer} corresponding to the {@link IClasspathEntry} if one exists, otherwise
+   * <code>null</code>
+   */
+  private LibraryClasspathContainer getLibraryClasspathContainer(IClasspathEntry entry) throws JavaModelException {
+    if (entry.getEntryKind() == IClasspathEntry.CPE_CONTAINER) {
+      IClasspathContainer container = JavaCore.getClasspathContainer(entry.getPath(), javaProject);
+      if (container instanceof LibraryClasspathContainer) {
+        return (LibraryClasspathContainer) container;
+      }
+    }
+    return null;
+  }
+
+  private void setSourceAttachmentPathForEntry(IJavaProject javaProject, IPath sourcePath, IProgressMonitor monitor,
+      LibraryClasspathContainer libraryContainer, IClasspathEntry[] classpathEntries, int j)
+      throws JavaModelException, IOException, CoreException {
```

Comment:
I agree, looks rather complex. The intent was to reduce the 5-6 level of indentation in `setSourceAttachmentPath`. I'll give it another try.